Adding specifications for specifying crystal structure atomic positions (#49)
Materials-Consortia/OPTiMaDe: optimade.md

  occurrence number in the **reduced** chemical formula and replace them with
  subsequent alphabet letters A, B, C and so on.

### <a name="h.6.2.5">6.2.5. dimension\_types</a>

* Description: a list of three integers. For each of the three 

>   directions indicated by the three lattice vectors (see property `lattice_vectors`) this list indicates if that direction is periodic (value 1) or non-periodic (value 0). Note: each periodicity direction is *not* one of the Cartesian x, y, z directions but the directions specified by the `lattice_vectors`.

Reviewer: Some very minor language editing, because, technically, the directions *could*, of course, be the Cartesian x,y,z directions, if that is where the lattice vectors point. 
```suggestion
  directions indicated by the three lattice vectors (see property `lattice_vectors`) this list indicates if that direction is periodic (value 1) or non-periodic (value 0). Note: the elements in this list each refer to the direction of the corresponding entry in `lattice_vectors` and *not* the Cartesian x, y, z directions.
```

* Description: Cartesian positions of each site. A site is 
  an atom, a site potentially occupied by an atom, or a placeholder 
  for a virtual mixture of atoms
  (e.g. in a virtual crystal approximation).
* Requirements: 
  * This property is REQUIRED.
  * It MUST be a list of length N times 3, where
    N is the number of sites in the structure.
  * An entry MAY have multiple sites at the same Cartesian position
    (for a relevant use of this, see, e.g., the `assemblies` property).
* Examples:
  * `[[0,0,0], [0,0,2]]` indicates a structure with two sites,
    one sitting at the origin and one along the (positive) `z` axis, 
    two angstrom away from the origin.

### <a name="h.6.2.8">6.2.8. species\_at\_sites</a>

* Description: Name of the species at each site (where values 
  for sites are specified with the same order of the 
  `cartesian_site_positions` property). The properties of 
  the species are found in the `species` property. 
* Requirements: 
  * This property is REQUIRED.
  * It MUST be a list of strings. It MUST have 
    length equal to the number of sites in the structure (first dimension of the `cartesian_site_positions` list). 
  * Each string MUST be a 

>     valid key of the dictionary specified by the `species` property.

Reviewer: In line with my longer comment: to make it possible for future extensions of the spec to give reserved names in a separate namespace special meaning (e.g., ":C" or "6" for standard Carbon). Also, it may be nice to not force clients to handle overly weird species names, e.g., with non-printable unicode in them. 
```suggestion
    valid key of the dictionary specified by the `species` property. The requirements on this string are the same as for property names, i.e., it can be of any length, may use upper and lower case letters, the underscore, and the digits 0-9, but cannot begin with a digit.    
```

Author: The only comment I have here is if the DB internally uses "unallowed" characters, then we are forcing it to rename all sites. 

**Suggestion**: Maybe a safe enough option is to just limit our namespace, e.g. simply forbid names to start with a character (e.g. `:`)? 

In the unlikely case that a DB is already doing it, it could automatically do some sort of escaping (also to make sure that when reading the name back, it can convert it back to the internal representation). 

E.g. (this is just an example, it's client specific and will not go in the specs, just for the discussion here):
- use an escape character, say `~`
- prepend `~` if the first character is `:` or the first character is `~`

so e.g.
- "A" -> "A"
- ":B" -> "~:B"
- "~C" -> "~~C"
that guarantees that we don't make it impossible to store names with different characters, nor to make round trips (DB->Optimade->DB)

Reviewer: But don't we still want to avoid forcing clients to handle names with 'very unexpected characters', e.g., non-printable unicode? (Say, the right-to-left mark...) Species names are a kind of 'variable names', and for such it is common to limit what is allowed.

I think it actually would be safe to let implementations invent any scheme of their own to 'downconvert' identifiers into the allowed character set. But, sure, if we really want roundtrips to be unicode-safe, what about saying that any character outside [a-zA-Z_][a-zA-Z_0-9]* needs to be escaped using, e.g., javascript notation; e.g., \xA9 is the copyright symbol, \x5C is a backslash, and a right-to-left mark would be \u200F. A client that want to worry about these things will thus interpret "Ti\xA9" as Ti©. But, a client that does not care will do fine by just regarding "Ti\xA9" as an abstract species identifier.

>     valid key of the dictionary specified by the `species` property.

Reviewer: Perhaps like so:
```suggestion
    valid key of the dictionary specified by the `species` property. This string can be of any length, may use upper and lower case letters, the underscore, the hyphen -, the colon :, the period ., and the digits 0-9. While it is recommended to keep species identifiers within these limitations, characters outside the allowed set may be encoded using the backslash symbol \. A sequence \uABCD where ABCD is a four digit hexadecimal number refers to the unicode character with this id. To encode an explicit backslash, use \u005C. At this point, no other backslash constructs are allowed beyond those on the form \uABCD. 
```
If you feel strongly that the second sentence should be expanded with more characters, feel free to suggest that. I am really stretching my imagination to see what this is meant to be used for. Note: if we adopt this, a notation for "special species" should stay outside what is allowed.

  occurrence number in the **reduced** chemical formula and replace them with
  subsequent alphabet letters A, B, C and so on.

### <a name="h.6.2.5">6.2.5. dimension\_types</a>

* Description: a list of three integers. For each of the three 
  directions indicated by the three lattice vectors (see property `lattice_vectors`) this list indicates if that direction is periodic (value 1) or non-periodic (value 0).
* Requirements: 
  * This property is required.
  * it must be a list of length 3.
  * each element must be an integer and can assume only the value 0 or 1.
* Examples:
  * For a molecule: `[0, 0, 0]`
  * For a wire along the direction specified by the third lattice vector: `[0, 0, 1]`
  * For a 2D surface/slab, periodic on the plane defined by the first and third
    lattice vector: `[1, 0, 1]`
  * For a bulk 3D system: `[1, 1, 1]`

### <a name="h.6.2.6">6.2.6. lattice\_vectors</a>

* Description: The three lattice vectors in Cartesian coordinates, in angstrom
* Requirements/convention: 
  * This property is required, except when `dimension_types` is equal to `[0, 0, 0]` (in this case it is optional).
  * It must be a 3 times 3 list of lists, each element being a float.

>   * The fastest index runs over the x, y, z Cartesian coordinates, while

Author: Ok thanks. It's common when speaking of matrices in memory, but indeed it might not be clear in this context (even if since we have a list of lists, it should be clear). We have examples so I'm merging this.
